### DISCUSSION TITLE
Library questions

### DISCUSSION THREAD

**Discussion by mhw0** [mhw0/libethc#10] [Q&A]
Ask anything about the library.

**Comment by DerXanRam** [comment]
Hello bro ... as u mention in the [doc](https://mhw0.github.io/libethc/abi.html), `eth_abi` can also use to decode. Does it decode ABI responses, like when u call `eth_call ` RPC function, it reads contract without spending real gas coast and resturns the ABI function response. Does it decodes this? Thanks alot :pray:

**Reply by mhw0** [reply]
Hey, yes, it can. But, you should know the exact structure of the ABI. Take a look at this test:

https://github.com/mhw0/libethc/blob/b39426fc867d1730270e8164a3145ba26bcd057b/test/test-eth-abi.c#L23-L27

We're creating a new `eth_abi` struct from the hex string, and we exactly know that the ABI is going to have 2 members. Thus, you can easily decode the encoded ABI.

**Comment by DerXanRam** [comment]
sorry but It's not clear bro. let's say, the `eth_call` RPC responded the below data for the abi function 
`function allPairs(uint) external view returns (address pair);`

The returned data
`0x0000000000000000000000003139ffc91b99aa94da8a2dc13f1fc36f9bdc98ee`

Can u give me some code snippet like example please?

**Reply by mhw0** [reply]
When you initialize `eth_abi` struct using `eth_abi_from_hex`, it sets the mode to `ETH_ABI_DECODE`, and the functions like `eth_abi_uint8`, `eth_abi_address`, `eth_abi_bool` etc. will behave differently. They decode the data and write the results to variables you passed as arguments.

If the contract has this signature: `function allPairs(uint) external view returns (address pair)`, you see that it returns only one address, and you call `eth_abi_address` only once to get the decoded data. Here is an example (didn't test):

```c
struct eth_abi abi;
char *addr;
// initialize abi struct from encoded abi
eth_abi_from_hex(&abi, 
       "0x0000000000000000000000003139ffc91b99aa94da8a2dc13f1fc36f9bdc98ee", -1);
// decoding happens here
eth_abi_address(&abi, &addr);
// free the abi struct
eth_abi_free(&abi);
printf("decoded address: %s\n", addr);
// free the address
free(addr); 
```

If your function returned address and bool, you would call `eth_abi_address` and `eth_abi_bool` to decode it.

**Reply by DerXanRam** [reply]
Now everything is crystal clear. Thanks alot :pray:

**Reply by DerXanRam** [reply]
just question.
some times there is a time abi functions return uint112 data type. Max data type found in the library is uint64_t. So does it hild it?

**Reply by mhw0** [reply]
We would need to use libgmp for that as there is no `uint112` in C.  Opened an issue: https://github.com/mhw0/libethc/issues/15

**Reply by DerXanRam** [reply]
> We would need to use libgmp for that as there is no `uint112` in C.  Opened an issue: https://github.com/mhw0/libethc/issues/15

Ok. Tnx man 🙏

**Comment by DerXanRam** [comment]
Hello bro.  I'm trying to make the RPC header file and json parser using c as u suggest to make the header compatible....  roughly, json parser seems easy but when i enter deep down things become complex. I suggest using premade libraries to save time and effort. what do u think?

**Reply by mhw0** [reply]
Hi. Uhm.. Okay. This one looks promising: https://github.com/json-c/json-c. If you're working on RPC calls, please consider this API:
```c
struct eth_rpc rpc; // rpc struct?
struct eth_rpc_call call; // stores the rpc call?
eth_rpc_init(&rpc, "https://rpc.node.abc", -1); // third one is the length of the url, -1 if NULL terminated

eth_rpc_call(&call, &rpc, /* jsonrpc body? string? struct? */, /* http headers? */); // this blocks the thread? do we need "async" version?

// call.status  http status? int?
// call.result  parsed jsonrpc "result" field? struct?
// call.id  parsed jsonrpc id field? int?

eth_rpc_call_free(&call); // free the call?
eth_rpc_free(&rpc); // free the rpc?
```
`rpc.h` file should be in `include/ethc/` and `rpc.c` in `src/` so we can easily import it like this: `#include <ethc/rpc.h>`

We also have batch and notification requests :)

**Reply by mhw0** [reply]
And, let's discuss it in here: https://github.com/mhw0/libethc/pull/13

**Comment by DerXanRam** [comment]
is there any compatible data type in `eth_abi_` functions to encode and/or decode string ?

**Reply by mhw0** [reply]
`eth_abi_bytes` should do the trick

**Reply by DerXanRam** [reply]
can it contains 20-25 length string?

**Reply by mhw0** [reply]
it's not fixed, so, even 300-400+

**Comment by DerXanRam** [comment]
> `eth_abi_bytes` should do the trick



> it's not fixed, so, even 300-400+

wow....nice...tnx very much for ur kind and fast response :pray:

**Comment by DerXanRam** [comment]
Hello bro. I have this data type
```
mpz_t amt;
mpz_init(amt);
```

for `eth_abi_mpint(&abi,amt);` but i can not find a way to copy a `char*` content to `amt` ? I need a hand pls :pray:  The content in `char*` is only numbers.

**Reply by mhw0** [reply]
Hello. You can use functions like [`mpz_init_set_str`](https://gmplib.org/manual/Simultaneous-Integer-Init-_0026-Assign) to initialize `mpz`:
```c
mpz_init_set_str(&amt, "0xfffff", 0);
```

**Comment by DerXanRam** [comment]
> Hello. You can use functions like [`mpz_init_set_str`](https://gmplib.org/manual/Simultaneous-Integer-Init-_0026-Assign) to initialize `mpz`:
> 
> ```c
> mpz_init_set_str(&amt, "0xfffff", 0);
> ```

I tried it but the result is not correct
*char content is `16171028259755471471`
content inside amt(printed using gmp_printf("%Zx\n", amt) `e06b08864387326f`

**Reply by mhw0** [reply]
`16171028259755471471` is `e06b08864387326f` (in hexadecimal format). Try printing it like this: `gmp_printf("%Zd\n", amt);`

**Reply by DerXanRam** [reply]
not working mate.....

**Comment by DerXanRam** [comment]
> gmp_printf("%Zd\n", amt);

It didn't work...
*char is  `16171028259755471471`
amt is `6833471685073593933929741`

and also the problem is not with the print. I want to pass the `amt` to `eth_abi_mpint(&abi,amt);` and the RPC responded with message `"execution reverted"` . So if the `amt` stores in hex format,  the RPC end will reject my request.

**Reply by mhw0** [reply]
Can you do `eth_abi_to_hex` and see if it's giving the correct result?

**Reply by mhw0** [reply]
And, send me the code please so I can test myself.

**Reply by DerXanRam** [reply]
```
string getAmountsOut(string amountIn, string pathA, string pathB, string exchangeAddress = UniswapexchangeAddress)
	{
		char *a, *b, *c;
		a = strdup(pathA.c_str());
		b = strdup(pathB.c_str());
		c = strdup(amountIn.c_str());
	
		cout<<"char is"<<c<<endl;
        mpz_t amt;
		mpz_init(amt);
		mpz_init_set_str(amt,c,0);
		cout<<"amt is ";
		gmp_printf("%Zd\n", amt);
        cout<<endl;
		struct eth_abi abi;
		char *fn = "getAmountsOut(uint256,address[])", *hex;
		uint64_t arraySize = 2;
		size_t hexlen;
		eth_abi_init(&abi, ETH_ABI_ENCODE);
		eth_abi_call(&abi, &fn, NULL);
		eth_abi_mpint(&abi,amt);
		eth_abi_array(&abi, &arraySize);
		eth_abi_address(&abi, &a);
		eth_abi_address(&abi, &b);
		eth_abi_array_end(&abi);
		eth_abi_call_end(&abi);

		ok(eth_abi_to_hex(&abi, &hex, &hexlen));
		ok(eth_abi_free(&abi));
		/*end of abi encoding*/

		json response = eth_call(hex, exchangeAddress); /*sending rpc call*/

		string result;
		//cout << "dump " << response.dump() << endl;
		if (response.contains("result"))
			result = response["result"];
		else
		{
			cout << response.dump() << endl;
			exit(-1);
		}

		/*copying ABI result to data*/
		char *data = strdup(result.c_str());
		// decoding
		struct eth_abi abi1;
		mpz_t r1, r2;
		mpz_init(r1);
		mpz_init(r2);

		uint64_t arrSize = 2;
		eth_abi_from_hex(&abi1, data, -1);
		eth_abi_array(&abi1, &arrSize);
		eth_abi_mpint(&abi1, r1);
		eth_abi_mpint(&abi1, r2);
		eth_abi_array_end(&abi1);
		eth_abi_free(&abi1);
		//gmp_printf("%Zd %s\n", r1, "test");

		char *temp;
		gmp_asprintf(&temp, "%Zd", r2);
		string t3=temp;
        free(temp);

		string decodedRes(t3);
		return decodedRes;
	}
```

**Reply by mhw0** [reply]
It looks correct, but you don't have to do `mpz_init(amt);` as `mpz_init_set_str` initializes the `mpz_t` itself. Can you remove the first one and try again?

**Comment by DerXanRam** [comment]
> It looks correct, but you don't have to do `mpz_init(amt);` as `mpz_init_set_str` initializes the `mpz_t` itself. Can you remove the first one and try again?

Didn't work mate........

**Reply by mhw0** [reply]
There is no way that if you initialize it from string (`16171028259755471471`) it gives you `6833471685073593933929741`. I tested the initialization part and it's working as intended.

**Reply by mhw0** [reply]
Also, let's keep all the comments in one question. No need to create another one.

**Reply by DerXanRam** [reply]
It prints correctly. The problem is , the data passed to eth_abi_mpint(&abi,amt); is still deformed. and the RPC sends this message "message":"execution reverted:

**Reply by DerXanRam** [reply]
and sorry :pray:  for my mess in the disscution section. i will correct it>

**Reply by DerXanRam** [reply]
wait ...

**Reply by DerXanRam** [reply]
now the RPC responded 
`"message":"execution reverted: ds-math-mul-overflow"`

**Reply by mhw0** [reply]
> now the RPC responded
> "message":"execution reverted: ds-math-mul-overflow"

Can you please send the ABI output?

**Reply by DerXanRam** [reply]
abi data is d06ca61fe165bae36fe1e4ad000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000020000000000000000000000006b175474e89094c44da98b954eedeac495271d0f000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2

**Reply by DerXanRam** [reply]
and 
char is16241593117178979501
amt is 16241593117178979501

**Reply by DerXanRam** [reply]
I think this `d06ca61fe165bae36fe1e4ad` part of the ABI is the `amt`. right? and the data is deformed as u see, when u convert it to decimal...

**Reply by mhw0** [reply]
I think I know what is causing this. Solidty ABI spec says that if the encoding element is `bytes`, it goes left to right: `ffffff000000000`, if it's an integer it will be the opposite(right to left): `000000000ffffff`. But, `eth_abi_mpint` uses `eth_abi_bytes32` for some reason (I don't remember why I did that).

**Reply by mhw0** [reply]
Opened an issue: #20

**Reply by mhw0** [reply]
> what shall i do bro.... i u see the number i want to send is 20 or 22 digit uint. But uint64_t max holds 19 digits as i observed. Is there any solution??

temporary solution would be to convert that `"16241593117178979501"` string to 32-bit byte array and use `eth_abi_bytes32`. For example if the number is `0xff` it would look like this:
```c
uint8_t bytes[32] = {0x00, /*<30 0xs>*/, 0xff};
eth_abi_bytes32(&abi, bytes);
```

**Comment by DerXanRam** [comment]
Can this library decode encoded abi data? i mean...
`0x38ed1739000000000000000000000000000000000000000000000000058d15e17628000000000000000000000000000000000000000000000000504560ab874a54b97a9800000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000e41137e6df7dece3491698f1e24ef6bcc2fbcb2000000000000000000000000000000000000000000000000000000000651c21180000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000004ed0791727029144f08b5c43e746a4a75026510d`

the above data is that sent using rawtransaction. the data contains this function and respective inputs
`swapExactTokensForTokens(uint256 amountIn, uint256 amountOutMin, address[] path, address to, uint256 deadline)`

so if i want to decode the raw data to function name and passed parameters, how do i accomplish that?

**Reply by mhw0** [reply]
Yes, you can. Just pass `ETH_ABI_DECODE` instead and the encoding part remains the same.

**Reply by DerXanRam** [reply]
wow... tnx very much :pray:

**Reply by DerXanRam** [reply]
My decode function runs but decoded outputs are "0"..  this is the code
```
void abiDecoder(string data)
	{
		char *input = strdup(data.c_str());
        char *fn="swapExactTokensForTokens(uint256,uint256,address[],address,uint256)";
		uint64_t amin;
		uint64_t amout;
        uint64_t arrlen;
		uint64_t deadline;
		int inputlen=1;
		char *add[8];
		char *addTo;
		// decoding
		struct eth_abi abi;
		uint8_t *bytes=0;
		uint64_t amt;
		size_t a; // result length;
		eth_abi_from_hex(&abi,input,523);
		eth_abi_init(&abi,ETH_ABI_DECODE);
		eth_abi_call(&abi, &fn, NULL);
		eth_abi_uint64(&abi,&amin);
		eth_abi_uint64(&abi,&amout);

		eth_abi_array(&abi,&arrlen);
		for(uint64_t i=0;i<arrlen;i++)
		    eth_abi_address(&abi,&add[i]);
		eth_abi_array_end(&abi);

		eth_abi_address(&abi,&addTo);
		eth_abi_uint64(&abi,&deadline);
 		eth_abi_call_end(&abi);
		//ok(eth_abi_to_hex(&abi, &hex, &hexlen));
		ok(eth_abi_free(&abi));

		char *final = (char *)bytes;
		cout << "decoded to address is " <<addTo << endl;
	}
```

outtput is 

`decoded to address is 0000000000000000000000000000000000000000`

the data i passed is 
`0x38ed1739000000000000000000000000000000000000000000000000058d15e17628000000000000000000000000000000000000000000000000504560ab874a54b97a9800000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000e41137e6df7dece3491698f1e24ef6bcc2fbcb2000000000000000000000000000000000000000000000000000000000651c21180000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000004ed0791727029144f08b5c43e746a4a75026510d`

**Reply by mhw0** [reply]
other values? are they also zero?

**Reply by DerXanRam** [reply]
Very sorry for my late reply... yeaa other values are also zero

**Reply by mhw0** [reply]
Oh, I think you should remove `eth_abi_init` (`eth_abi_from_hex` already initializes the `abi` struct)

**Reply by DerXanRam** [reply]
If i remove `eth_abi_init` and i do `eth_abi_from_hex(&abi,input,data.length()) it worked :+1: ... tnx a lot :pray:

**Comment by DerXanRam** [comment]
Hello brother. As u can see in [this example](https://github.com/mhw0/libethc/discussions/35#discussioncomment-7282167) of signing raw tnx, i use uint8_t for `value`. But some small and new tokens, that have 18 decimals and I have to move like in 100k-2M and `value` becomes bigger.  Now here is my question, when i convert this type of number to hex, it becomes more than 8 bits (like 10 and more) and uint8_t can not hold it. So for this type of amount, can i use uint32_t?

**Reply by mhw0** [reply]
Hello, Yes, you can use any type for that field.

**Reply by DerXanRam** [reply]
ok thanks bro :pray:

**Comment by DerXanRam** [comment]
Hello bro. How are u. can u look my comment in this [closed issue ](https://github.com/mhw0/libethc/issues/7#issue-1855641858) please?

**Comment by DerXanRam** [comment]
Hello bro. Is there a way to get `from` address(the address that signs the tx) using this info and libethc?

```
{
  "result": {
    "type": "0x0",
    "nonce": "0x2",
    "gasPrice": "0xb2d05e00",
    "maxPriorityFeePerGas": null,
    "maxFeePerGas": null,
    "gas": "0x5208",
    "value": "0x10b388db16d000",
    "input": "0x",
    "v": "0x93",
    "r": "0xf152db88c83765803a42b85e348e7b7335d87da193c5e6c6eecbebe6309d3604",
    "s": "0x8ae7a835aec59df3f171f178db4f4d031390f6ff5a99a7e32c19785e3b93cfc",
    "to": "0xbfe050d085de276ab759938c96e75ae4154a8096",
    "hash": "0x41bd78e14dca59e633e53ad4b78da4a6744286120a368be6d5dc2eb9b49a3070"
  }
}
```

**Reply by mhw0** [reply]
Hello, sorry for the late response. I think you can recover your public key (your address) from `v`, `r` and `s` signatures.

**Reply by DerXanRam** [reply]
No worries brother.....So to recover from r, s, and v, Should i use rlp in decode mode or account functions(i never use them before)?

**Reply by mhw0** [reply]
No, we need `secp256k1`'s recovery functions. I'll open an issue.

**Reply by mhw0** [reply]
https://github.com/mhw0/libethc/issues/37#issue-1984619712

**Comment by DerXanRam** [comment]
Hello bro. How are u. since last 2 months, this library is getting attention on different forums and sites. But it is still cooled here. anyway how is everything?

**Reply by mhw0** [reply]
Hey, I’m great, thank you for asking. I’ve been just busy lately. So, yeah. But, I will not stop maintaining this library.

> this library is getting attention on different forums and sites

what sites & forums are we talking about?

**Reply by DerXanRam** [reply]
i see a many mentions in GitHub. Outside GitHub, these are the pages i remember

https://awesome-web3.com/
https://www.libhunt.com/topic/eip
https://ethereum.stackexchange.com/questions/128042/how-to-use-web3-with-c
https://ethereum.stackexchange.com/questions/57103/call-a-view-function-of-an-ethereum-smart-contract-using-a-program-written-in-c
https://www.reddit.com/r/ethereum/comments/ki5fe2/how_to_access_ethereum_transactions_through_c/

**Comment by qingfengzxr** [comment]
> Ask anything about the library.

Hello, brothers, how can i use this library to decode rlp? I can't find some example or test in project files. Can you give me a little example code? Wish you a happy life!

**Reply by mhw0** [reply]
Hi @qingfengzxr You gotta initiate `eth_rlp` with RLP data using `eth_rlp_from_hex` first. Try this:

```c
uint8_t d0, d1;
struct eth_rlp rlp;

ok(eth_rlp_from_hex(&rlp, "c381ac08", -1) == 1);

ok(eth_rlp_array(&rlp) == 1);
  ok(eth_rlp_uint8(&rlp, &d0) == 1);
  ok(eth_rlp_uint8(&rlp, &d1 == 1);
ok(eth_rlp_array_end(&rlp) == 1);
ok(eth_rlp_free(&rlp) == 1);

printf("d0: %d, d1: %d\n", d0, d1);
```

**Reply by mhw0** [reply]
https://github.com/mhw0/libethc/blob/dev/test/test-eth-rlp.c

**Reply by qingfengzxr** [reply]
Ok, I'll give it a try. Thanks for the guidance! @mhw0
